distributed/tests/test_scheduler.py

```diff
@@ -371,6 +372,67 @@ def __del__(self):
     assert max(Refcount.log) <= s.total_nthreads
 
 
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_forget_tasks_while_processing(c, s, a, b):
+    events = [Event() for _ in range(10)]
+
+    futures = c.map(Event.wait, events)
+    await events[0].set()
+    await futures[0]
+    await c.close()
+    assert not s.tasks
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)] * 3,
+    config={"distributed.scheduler.worker-saturation": 1.0},
+)
+async def test_queued_release_multiple_workers(c, s, *workers):
+    async with Client(s.address, asynchronous=True) as c2:
+        event = Event(client=c2)
+
+        rootish_threshold = s.total_nthreads * 2 + 1
+
+        first_batch = c.map(
+            lambda i: event.wait(),
+            range(rootish_threshold),
+            key=[f"first-{i}" for i in range(rootish_threshold)],
+        )
+        await async_wait_for(lambda: s.queued, 5)
+
+        second_batch = c2.map(
+            lambda i: event.wait(),
+            range(rootish_threshold),
+            key=[f"second-{i}" for i in range(rootish_threshold)],
+            fifo_timeout=0,
+        )
+        await async_wait_for(lambda: second_batch[0].key in s.tasks, 5)
+
+        # All of the second batch should be queued after the first batch
+        assert [ts.key for ts in s.queued.sorted()] == [
+            f.key
+            for f in itertools.chain(first_batch[s.total_nthreads :], second_batch)
+        ]
+
+        # Cancel the first batch.
+        # Use `Client.close` instead of `del first_batch` because deleting futures sends cancellation
+        # messages one at a time. We're testing here that when multiple workers have open slots, we don't
+        # recommend the same queued tasks for every worker, so we need a bulk cancellation operation.
+        await c.close()
+        del c, first_batch
+
+        await async_wait_for(lambda: len(s.tasks) == len(second_batch), 5)
+
+        # Second batch should move up the queue and start processing
+        assert len(s.queued) == len(second_batch) - s.total_nthreads, list(
+            s.queued.sorted()
+        )
+
+        await event.set()
+        await c2.gather(second_batch)
```

Comment:
This test feels a bit heavy weight considering that we have a couple of very simple reproducers, see https://github.com/dask/distributed/issues/7396#issue-1494040437

If you feel strongly about this test, fine, but please add the other two very simple reproducers as well. Regardless of all the intricate timings and queuing, etc. The reproducers there should be true regardless of what internals look like

Reply:
I added the reproducer from https://github.com/dask/distributed/issues/7396#issue-1494040437 as well (with an Event to avoid timing issues): https://github.com/dask/distributed/blob/79525791d3d710533571cd49aff450539c949ffe/distributed/tests/test_scheduler.py#L375-L383

This test is for https://github.com/dask/distributed/issues/7401, which is a different issue (and I think maybe the same as https://github.com/dask/distributed/issues/7398). `Client.close` is the only codepath that can cause `client_releases_keys` to be called with multiple keys, which is the case that breaks.

Reply:
Added a simple test with `restart` as well (fails on main, reproducer for https://github.com/dask/distributed/issues/7398): https://github.com/dask/distributed/blob/31975e5b44552688d586c12c694d9a4716378cf1/distributed/tests/test_scheduler.py#L386-L394